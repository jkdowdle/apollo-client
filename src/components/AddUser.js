import React from 'react'
import { graphql } from 'react-apollo'
import gql from 'graphql-tag'

import { usersFeedQuery, UsersQuery } from './AllUsers'

const AddUserMutation = gql`
  mutation AddUser($input: UserInput!) {
    addUser(input: $input) {
      id
      firstName
      age
    }
  }
`

export class AddUser extends React.Component {
  input = {
    firstName: '',
    age: ''
  }

  updateForm = event => {
    const target = event.target
    const name = target.name
    const value = target.value

    this.input[name] = value
  }

  render() {
    const props = this.props
    return (
      <div>
        <h2>Add User</h2>
        <input type="text" name="firstName" onChange={this.updateForm} />
        <br />
        <br />
        {<input type="number" name="age" onChange={this.updateForm} />}
        <br />
        <br />
        <button onClick={props.addUser.bind(null, this.input)}>Add</button>
      </div>
    )
  }
}

const update = (store, { data: { addUser } }) => {
  const data = store.readQuery({ query: UsersQuery, variables: { cursor: null } })
  data.users.push(addUser)
  store.writeQuery({ query: UsersQuery, data })

  // const data = store.readQuery({ query: UsersQuery, variables: { cursor: null } })
  // data.users.unshift(addUser)

  // store.writeQuery({ query: UsersQuery, data })
}

const optimisticResponse = (input) => ({
    __typename: 'Mutation',
    addUser: {
      __typename: 'User',
      id: -1,
      optimistic: true,
      ...input
    }
})

const props = ({ mutate }) => ({
  addUser: (input) => mutate({ 
    variables: { input }, 
    update, 
    optimisticResponse: optimisticResponse(input) 
  })
})

export default graphql(AddUserMutation, {
  props
})(AddUser)
