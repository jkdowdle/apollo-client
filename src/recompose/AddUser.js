import React from 'react'
import { withState, withHandlers, pure, compose } from 'recompose'
import gql from 'graphql-tag'
import { graphql } from 'react-apollo'

import { usersList } from './AllUsers'

const defaultState = { firstName: '', age: '' }

const form = withState('form', 'setForm', defaultState)
const updateForm = ({ form, setForm }) => ({ target: { name, value }}) => setForm({ ...form, [name]: value })

const addUser = ({ addUser, form, setForm }) => (event) => {
  event.preventDefault()
  return addUser({ ...form })
    .then((res) => setForm({ firstName: '', age: '' }))
}

const handlers = withHandlers({
  updateForm,
  addUser
})

const AddUserPure = ({ updateForm, form: { firstName, age }, addUser, ...props }) => {
  return (
    <form onSubmit={addUser}>
      <h2>New User</h2>

      <input
        placeholder="User's name"
        type="text"
        name="firstName"
        value={firstName}
        onChange={updateForm}
      />

      <br />
      <br />

      <input
        type="number"
        name="age"
        value={age}
        onChange={updateForm}
      />

      <button type="submit">Go</button> 

    </form>
  )
}

const addUserMutation = gql`
  mutation AddUser($input: UserInput!) {
    addUser(input: $input) {
      id
      firstName
      age
    }
  }
`

const update = (store, { data: { addUser } }) => {
  const data = store.readQuery({ query: usersList, variables: {} })
  data.users.unshift(addUser)
  store.writeQuery({ query: usersList, data })
}

const optimisticResponse = (input) => ({
  __typename: 'Mutation',
  addUser: {
    __typename: 'User',
    id: -1,
    optimistic: true,
    ...input
  }
})

const props = ({ mutate }) => ({
  addUser: (input) => mutate({ 
    variables: { input }, 
    update, 
    optimisticResponse: optimisticResponse(input) 
  })
})

const mutation = graphql(
  addUserMutation,
  { props }
)

const AddUser = compose(
  mutation,
  form,
  handlers,
  pure
)(AddUserPure)

export default AddUser