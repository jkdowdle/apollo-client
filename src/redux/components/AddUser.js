import React from 'react'

export const AddUser = ({ handleSubmit, handleChange, firstName, age }) => {
  return (
    <form onSubmit={handleSubmit}>
      <div>
        <input type="text" value={firstName} name="firstName" onChange={handleChange} />
      </div>
      <div>
        <input type="text" value={age} name="age" onChange={handleChange} />
      </div>
      <div>
        <button type="submit">AddUSer</button>
      </div>
    </form>
  )
} 

export default AddUser