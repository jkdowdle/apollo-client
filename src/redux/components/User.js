import React from 'react'

export const User = ({ firstName, age }) => {
  return (
    <p>{firstName} - {age}</p>
  )
}

export default User