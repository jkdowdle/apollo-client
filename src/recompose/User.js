import React from 'react'
import { compose, branch, renderComponent, withState, withHandlers } from 'recompose'
import { graphql } from 'react-apollo'
import gql from 'graphql-tag'

import { usersList } from './AllUsers'

const edit = withState('edit', 'setEdit', false)

const toggleEdit = ({ setEdit, edit }) => (event) => {
  setEdit(!edit)
}

const handlers = withHandlers({
  toggleEdit
})

const input = withState('input', 'setInput', ({ firstName, age }) => ({ firstName, age }))
const updateInput = ({ input, setInput }) => ({ target: { name, value }}) => setInput({ ...input, [name]: value })

const updateUser = ({ updateUser, input, id }) => () => {
  // console.log('updating...', avail, input)
  return updateUser({ ...input, id })
} 

const handlersTwo = withHandlers({
  updateInput,
  updateUser
})

export const EditUserComponent = ({ toggleEdit, updateInput, firstName, deleteUser, updateUser, ...props }) => {
  console.log()
  return (
    <div>
      <input type="text" name="firstName" onChange={updateInput} value={firstName} />
       - <button onClick={updateUser}>Update</button>
       - <button onClick={toggleEdit}>Cancel</button>
       - <button onClick={deleteUser}>Delete User</button>
    </div>
  )
}

const EditUser = compose(
  input,
  handlersTwo
)(EditUserComponent)

const displayEditing = branch(
  ({ edit }) => edit,
  renderComponent(EditUser)
)

export const UserComponent = ({ firstName, age, toggleEdit }) => {
  return (
    <li>
      {firstName} - {age} - <button onClick={toggleEdit}>Edit</button>
    </li>
  )
}

export const deleteUserMutation = gql`
  mutation DeleteUser($input: Id!) {
    deleteUser(input: $input) {
      id
      firstName
      age
    }
  }
`

const optimisticResponse = (input) => ({
  __typename: 'Mutation',
  deleteUser: {
    ...input
  }
})

const update = (proxy, { data: { deleteUser } }) => {
  const data = proxy.readQuery({ query: usersList })
  const index = data.users.findIndex((user) => user.id === deleteUser.id)
  data.users.splice(index, 1)
  proxy.writeQuery({ query: usersList, data })
}

const props = ({ mutate, ownProps: { id, ...data }}) => {
  return {
    deleteUser() {
      mutate({ 
        variables: { input: { id } },
        update,
        optimisticResponse: optimisticResponse({ id, ...data })
      })
    }
  }
}

const mutation = graphql(
  deleteUserMutation,
  { props }
)

export const updateUserMutation = gql`
  mutation UpdateUser($id: Int!, $input: UserInput) {
    updateUser(id: $id, input: $input) {
      id
      firstName
      age
    }
  }
`

const propsTwo = ({ mutate }) => {
  return {
    updateUser() {
      mutate({
        variables: { id: 2, input: { firstName: 'Emily Ann', age: 27 } }
      })
    }
  }
}

const mutationTwo = graphql(
  updateUserMutation,
  { props: propsTwo }
)

const User = compose(
  edit,
  handlers,
  mutation,
  mutationTwo,
  displayEditing,
)(UserComponent)

export default User