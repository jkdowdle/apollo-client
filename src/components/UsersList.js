import React from 'react'
import { graphql } from 'react-apollo'
import gql from 'graphql-tag'

import AddUser from './AddUser'
import User from './User'
import SearchUsers from './SearchUsers'

const SearchUsersQuery = gql`
  query SearchUsersQuery($input: SearchUser) {
    searchUsers(input: $input) {
      id
      firstName
      age
    }
  }
`

export const UsersQuery = gql`
  query UsersQuery {
    users {
      id
      firstName
      age
    }
  }
`

export const usersFeedQuery = gql`
  query UsersFeedQuery($cursor: String) {
    usersFeed(cursor: $cursor) @connection(key: "usersFeed") {
      cursor
      users {
        id
        firstName
        age
      }
    }
  }
`

export const UsersList = ({ data, ...rest }) => {

  if (data.loading) {
    return 'loading'
  }

  // console.log('data', data)

  const users = data.users || data.searchUsers || []

  return (
    <div>
      <h1>Users List</h1>
      <AddUser />

      <SearchUsers />
      
      <hr />
      {users.map(({ id, ...props }) => <User key={id} {...props} id={id} />)}
    </div>
  )
}
/* <button onClick={rest.loadOlder}>Fetch More</button> */

const options = (props) => ({
  variables: { input: { searchTerm: 'a' } },
  // errorPolicy: 'all'
})

export const props = ({ data, ...rest }) => {
  return {
    data,
    loadOlder() {
      return data.fetchMore({
        variables: {
          cursor: data.usersFeed.cursor
        },
        updateQuery(previousResult, { fetchMoreResult: { usersFeed } }) {
          const { usersFeed: { users: previous }} = previousResult
          const { users: response } = usersFeed

          return {
            usersFeed: {
              ...usersFeed,
              users: [
                // ...previous,
                ...response,
              ]
            }
          }
        }
      })
    }
  }
}

export default graphql(UsersQuery, {
  options,
  props
})(UsersList)