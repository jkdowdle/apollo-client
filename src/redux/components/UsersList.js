import React from 'react'

import User from './User'

export const UsersList = ({ data: { users, searchUsers } }) => searchUsers.map((user) => <User key={user.id} {...user} />)

export default UsersList