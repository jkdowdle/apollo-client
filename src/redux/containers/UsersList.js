import React from 'react'
import { compose } from 'recompose'
import { graphql } from 'react-apollo'

import USERS_QUERY from '../graphql/UsersQuery'


import UsersList from '../components/UsersList'

const data = graphql()

export default compose(

)