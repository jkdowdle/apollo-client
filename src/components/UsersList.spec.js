import React from 'react'
import { render, mount, shallow } from 'enzyme'
import renderer from 'react-test-renderer'

import { UsersList, props as mapProps } from './UsersList'
import { User } from './User'
import { AddUser } from './AddUser'

describe('UsersList component', () => {
  const data = {
    users: [
      { id: 0, firstName: 'test1', age: 1 },
      { id: 1, firstName: 'test2', age: 2 }
    ]
  }

  it('should have 2 User components', () => {
    const wrapper = mount(<UsersList data={data} />)
    expect(wrapper.find(User).length).toBe(2)
  })

  it('should have an Adduser Component', () => {
    const wrapper = mount(<UsersList data={data} />)
    expect(wrapper.find(AddUser).length).toBe(1)
  })

  // it('should receive the correct props from graphql HOC', () => {
  //   const props = { data: { users: [] }, fetchMore: jest.fn() }
    
  //   const mapedProps = mapProps(props)
  //   const acctualProps = { data: { users: [] }, loadOlder: () => ({}) }

  //   console.log('retunred', mapedProps)

  //   expect(mapedProps).toEqual(acctualProps)
  // })

  it('should render UsersList correctly', () => {
    const tree = shallow(<UsersList data={data} />)

    expect(tree).toMatchSnapshot()
  })

})

