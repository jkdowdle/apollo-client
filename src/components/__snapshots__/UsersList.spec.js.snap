// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UsersList component should render UsersList correctly 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <UsersList
    data={
        Object {
            "users": Array [
              Object {
                "age": 1,
                "firstName": "test1",
                "id": 0,
              },
              Object {
                "age": 2,
                "firstName": "test2",
                "id": 1,
              },
            ],
          }
    }
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <h1>
          Users List
</h1>,
        <Apollo(AddUser) />,
        <hr />,
        Array [
          <Unknown
            age={1}
            firstName="test1"
/>,
          <Unknown
            age={2}
            firstName="test2"
/>,
        ],
        <button
          onClick={undefined}
>
          Fetch More
</button>,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "Users List",
        },
        "ref": null,
        "rendered": "Users List",
        "type": "h1",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {},
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {},
        "ref": null,
        "rendered": null,
        "type": "hr",
      },
      Object {
        "instance": null,
        "key": "0",
        "nodeType": "function",
        "props": Object {
          "age": 1,
          "firstName": "test1",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": "1",
        "nodeType": "function",
        "props": Object {
          "age": 2,
          "firstName": "test2",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "Fetch More",
          "onClick": undefined,
        },
        "ref": null,
        "rendered": "Fetch More",
        "type": "button",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <h1>
            Users List
</h1>,
          <Apollo(AddUser) />,
          <hr />,
          Array [
            <Unknown
              age={1}
              firstName="test1"
/>,
            <Unknown
              age={2}
              firstName="test2"
/>,
          ],
          <button
            onClick={undefined}
>
            Fetch More
</button>,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Users List",
          },
          "ref": null,
          "rendered": "Users List",
          "type": "h1",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {},
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {},
          "ref": null,
          "rendered": null,
          "type": "hr",
        },
        Object {
          "instance": null,
          "key": "0",
          "nodeType": "function",
          "props": Object {
            "age": 1,
            "firstName": "test1",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": "1",
          "nodeType": "function",
          "props": Object {
            "age": 2,
            "firstName": "test2",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Fetch More",
            "onClick": undefined,
          },
          "ref": null,
          "rendered": "Fetch More",
          "type": "button",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
